src/main/java/com/hotels/service/tracing/zipkintohaystack/forwarders/haystack/HaystackDomainConverter.java

```diff
@@ -24,23 +24,62 @@
 import java.util.Optional;
 import java.util.function.Consumer;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
 import com.expedia.open.tracing.Span;
 import com.expedia.open.tracing.Tag;
 
 /**
  * Converter between {@code Zipkin} and {@code Haystack} domains.
  */
+@Component
 public class HaystackDomainConverter {
 
     private static final String HAYSTACK_TAG_KEY_FOR_DATACENTER = "X-HAYSTACK-INFRASTRUCTURE-PROVIDER";
 
-    private HaystackDomainConverter() {
+    private final Logger logger = LoggerFactory.getLogger(HaystackDomainConverter.class);
+    private final boolean acceptNullTimestamps;
+    private final int maxTimestampDriftSeconds;
+
+    public HaystackDomainConverter(@Value("${pitchfork.validators.accept-null-timestamps}") boolean acceptNullTimestamps,
+                                   @Value("${pitchfork.validators.max-timestamp-drift-seconds}") int maxTimestampDriftSeconds) {
+        this.acceptNullTimestamps = acceptNullTimestamps;
+        this.maxTimestampDriftSeconds = maxTimestampDriftSeconds;
     }
 
     /**
      * Accepts a span in {@code Zipkin V2} format and returns a span in {@code Haystack} format.
+     * Returns {@link Optional#empty()} if the span is null.
      */
-    public static Span fromZipkinV2(zipkin2.Span zipkin) {
+    public Optional<Span> fromZipkinV2(zipkin2.Span zipkin) {
+        // TODO: move this outside of the converter and into the ZipkinController.addSpans chain
```

Comment:
one side-effect of dropping is that you can lose a parent for a trace in progress. spans are usually reported depth first. you might consider dropping the trace instead or something else.. up to you

Reply:
Thanks for the feedback, it's always much appreciated!

I'll reply to both comments here as they're related.

This is related to this ticket: https://github.com/ExpediaDotCom/haystack/issues/611
Still not sure yet but it looks like the performance problems with the queries to Elasticsearch (aggregations) are a result of bad timestamps.

Based on the data we have so far there's reason to believe that some rogue service is using a wrong timestamp (possibly reporting in millis and not micros or with a timestamp that someone hardcoded). (I don't think the null timestamps are causing any issue in Haystack but I figured it was super cheap to add that optional validation as well).

Regarding the issues like orphan spans, etc. I guess it depends on how you look at this feature. I see it more of a quick way to troubleshoot a problem and stop offenders and not so much a fully fledged feature to sanitize the data and handle those edge cases.